win32ss/user/user32/windows/cursoricon.c

```diff
@@ -2041,16 +2041,53 @@ HANDLE WINAPI CopyImage(
     switch(uType)
     {
         case IMAGE_BITMAP:
+            if (!hImage)
+            {
+                SetLastError(ERROR_INVALID_HANDLE);
+                break;
+            }
             return BITMAP_CopyImage(hImage, cxDesired, cyDesired, fuFlags);
         case IMAGE_CURSOR:
         case IMAGE_ICON:
-        /* HACK: Copying bitmaps with LR_COPYFROMRESOURCE flag fails. CORE-17902.
-         * This is a way to return the original bit map if we need
-         * the icons to show up. We need a simpler test. */
         {
-            HANDLE handle = CURSORICON_CopyImage(hImage, uType == IMAGE_ICON, cxDesired, cyDesired, fuFlags);
-            if (!handle && (fuFlags & (LR_COPYFROMRESOURCE|LR_COPYRETURNORG)))
-                handle = CURSORICON_CopyImage(hImage, uType == IMAGE_ICON, cxDesired, cyDesired, (fuFlags & ~LR_COPYFROMRESOURCE));
+            HANDLE handle;
+            if (!hImage)
+            {
+                SetLastError(ERROR_INVALID_CURSOR_HANDLE);
+                break;
+            }
+            handle = CURSORICON_CopyImage(hImage, uType == IMAGE_ICON, cxDesired, cyDesired, fuFlags);
+            if (!handle && (fuFlags & LR_COPYFROMRESOURCE))
+            {
+                /* Test if the hImage is the same size as what we want by getting
+                 * its BITMAPINFO and comparing it to the desired size. */
+                HDC hDC;
+                ICONINFO iconinfo = { 0 };
+                /* struct is for BITMAPINFO plus a full-sized color table */
+                struct
+                {
+                    BITMAPINFOHEADER bmiHeader;
+                    RGBQUAD bmiColors[256];
+                } bmi;
+
+                if (!GetIconInfo((HICON)hImage, &iconinfo))
+                    ERR("GetIconInfo Failed. hImage %p\n", hImage);
```

Comment:
should we continue if this call fails? (the iconinfo.hbmMask and hbmColor below would remain NULL in that case)

Reply:
I have revised the code so that on an error here, we do ```return NULL;```. Thanks